### ISSUE TITLE
local variables out of scope in debug

### ISSUE THREAD

**Issue by Papirosnik** [pkulchenko/ZeroBraneStudio#658]
zbs from today's master branch (mobdebug 0.6371), gideros Lua interpreter win7 64bit...
All seems works  fine but breakpoint at last _end_ in a scope Is worthless for debugging purposes.
In the screenshot below I intended to get a value of  local variable _a_ and, as you can see it is already out of scope.
![ss1](https://cloud.githubusercontent.com/assets/1651549/16071646/2f287f4c-32e5-11e6-935a-4d973df073e8.JPG)

 Sure, I can add a fake line with useless content (something like _local b_) and set a breakpoint to it but this is pretty inconvenient.
![ss2](https://cloud.githubusercontent.com/assets/1651549/16071659/4066971c-32e5-11e6-9607-46f9384005e0.JPG)

 It would be nice, if debugger could stop on breakpoint at line with end  before performing 'end-of-scope' operation...

**Comment by pkulchenko** [comment]
Hm, I'm not sure if Gideros does something differently, but it works with the interpreters included with the IDE (I checked the default LuaJIT and Lua 5.2 on Windows 8):

![zerobrane-studio-end-debugger-stop](https://cloud.githubusercontent.com/assets/77071/16084040/0f820028-32cc-11e6-8356-66f998757fb4.png)

Assuming the `end` in your screenshot is for a function, it should work (at least from the debugger perspective).

**Comment by Papirosnik** [comment]
However the problem exists.
I managed to reproduce it with pure lua too.
And the problem is because of false breakpoint triggering at last _end_ of function.
I wrote a simple case to demonstrate it:
```
local f = function()
	local a = 1
	print(a)
end
f()
```
Set a breakpoint at end of function. F5.
![ss1](https://cloud.githubusercontent.com/assets/1651549/16091749/b52eccec-333d-11e6-835d-2afe149910d8.PNG)

Breakpoint is triggered in main chunk! Without invoking local function! And sure, debuger knows nothing about local '_a_' (because it is stopped in global scope)

Press F10:
![ss2](https://cloud.githubusercontent.com/assets/1651549/16091796/e458a98e-333d-11e6-989b-21a43f1dcfd8.PNG)
Only now we are going to call our function. Press F5 once again.
Now all looks as expected. Our breakpoint is triggered in local scope and debuger sees the local variable.
![ss3](https://cloud.githubusercontent.com/assets/1651549/16091833/0bba6666-333e-11e6-929a-1e9d94227c37.PNG)

**Comment by pkulchenko** [comment]
> Breakpoint is triggered in main chunk! Without invoking local function! And sure, debuger knows nothing about local 'a' (because it is stopped in global scope)

Right, but it's not an issue; the first time the breakpoint is triggered on line 4 is because that's the end of function declaration (assignment to a local variable `f`) and since the function is not executed yet, the local variable `a` doesn't exist (as you correctly pointed out the control is in the `main` chunk).

The second time it's stopped on line 4 is when the function is executed right before leaving it, at which point `a` value is already set and is available in Stack and Watch panels.

**Comment by Papirosnik** [comment]
Yeah, you are right. When I use global function declaration (instead of local var assignment), the breakpoint doesn't trigger twice. Only once upon function invoking.
Initially I was confused by OOP-aprocah: function MyClass:foo() end
In fact, in this case function is also assigned to some variable under hood...
Thanks for the clarification!
